### PR TITLE
Only run codesign if there have been substitutions

### LINKED ISSUE #6226: 3.5.0-alpha1: Can't build without `codesign` on macOS
## Expected Behavior

Able to build Dune 3.5.0-alpha1 on macOS + Nixpkgs

## Actual Behavior

```
dune> File "_unknown_", line 1, characters 0-0:
dune> Error: Program codesign not found in the tree or in PATH
dune> Hint: codesign should be part of the macOS installation
```

I tried adding [`darwin.sigtool`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/sigtool/default.nix) but I believe it brings up a password prompt that nix doesn't like.

## Reproduction

- PR with a reproducing test: 

1.
1.
1.

## Specifications

- Version of `dune` (output of `dune --version`):
- Version of `ocaml` (output of `ocamlc --version`)
- Operating system (distribution and version):

## Additional information 

- Link to gist with verbose output (run `dune` with the `--verbose` flag):

### PR DESCRIPTION
**Pull request by emillon**
This ensures that we're not running `codesign` in cases we don't strictly need it. This in turn prevents a regression in macos+nix, where the codesign binary is not in PATH.

Closes #6226

### CODE REVIEW

**Comment by emillon** [comment]
@anmonteiro can you try this patch to see if this works on mac+nix? :pray:

File: src/dune_rules/artifact_substitution.mli [hunk 1]
```diff
@@ -72,13 +72,15 @@ val copy_file :
...
+    source file.
+
+    Return whether a substitution happened. *)
 val copy :
      conf:conf
   -> input_file:Path.t
   -> input:(Bytes.t -> int -> int -> int)
   -> output:(Bytes.t -> int -> int -> unit)
-  -> unit Fiber.t
+  -> bool Fiber.t
```
Comment (rgrinberg) [inline src/dune_rules/artifact_substitution.mli:83]: Can the return value be a variant of some sorts? Returning true or false is not very clear for the reader.

**Comment by rgrinberg** [comment]
Do nix builds skip the substitution step or something? Otherwise, I don't see how it fixes the original issue.

**Comment by emillon** [comment]
The reasoning is the following: the sign hook runs in all cases where `copy` is called. This includes cases where the binary does not contain anyplaceholders. Because of this, we're calling `codesign` in cases that are not necessary (the signature would still be valid).
This is the regression found in #6226: users with a M1 mac, even with nix, should have something that keeps working when they don't use substitution (this was working pre-3.5.0).
For M1 users that want to use substitution, there used to be no support pre 3.5; 3.5 brings support to them (except for nix users), and we'll have to find a way to make this work for nix users post 3.5 (maybe by tweaking nix?).

**Comment by anmonteiro** [comment]
If my issue is delaying a Dune 3.5 release, it should be OK to go ahead the current solution, if no one else is experiencing these issues. I can work around it on my end for the time being.

**Comment by rgrinberg** [comment]
> If my issue is delaying a Dune 3.5 release, it should be OK to go ahead the current solution, if no one else is experiencing these issues. I can work around it on my end for the time being.

Thanks for the offer, but we'd rather not release with known regressions.

@emillon is this just an issue of looking up `codesign` too eagerly?

**Comment by emillon** [comment]
Yes, we're looking for codesign a bit too early. I'll amend the PR.

**Comment by emillon** [comment]
@anmonteiro can you try this one? we're now only looking up `codesign` in `PATH` as we're substituting (if needed).

**Comment by anmonteiro** [comment]
The latest rev seems to work like before. Thank you!

**Comment by emillon** [comment]
Perfect! Thanks a lot for testing.

**Pull request merged**